### PR TITLE
TLSv1.3 cipher suites: fixes

### PR DESCRIPTION
**Pull request by SparkiDev**
# Description

Handle multiple instances of the same cipher suite being in the server's list.
Fix client order negotiation of cipher suite when doing pre-shared keys.

Fixes zd#14820

# Testing

Added tests to api.c: test_tls13_cipher_suites()

# Checklist

 - [ x added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

File: src/ssl.c [hunk 1]
```diff
@@ -18718,6 +18718,8 @@ size_t wolfSSL_get_client_random(const WOLFSSL* ssl, unsigned char* out,
         ssl->keys.encryptionOn = 0;
         XMEMSET(&ssl->msgsReceived, 0, sizeof(ssl->msgsReceived));
 
+        InitSSL_Suites(ssl);
```
Comment (philljj) [inline src/ssl.c:18721]: Does return value of InitSSL_Suites need to be checked?

**Review by douzzer — CHANGES_REQUESTED** [review]
had a clean run through multi-test.

File: src/ssl.c (hunk 1 above)
Comment (SparkiDev) [reply src/ssl.c:18721]: fixed

File: src/tls13.c [hunk 2]
```diff
@@ -4755,16 +4755,34 @@ static void RefineSuites(WOLFSSL* ssl, Suites* peerSuites)
 {
     byte   suites[WOLFSSL_MAX_SUITE_SZ];
     word16 suiteSz = 0;
-    word16 i, j;
+    word16 i;
+    word16 j;
 
     XMEMSET(suites, 0, WOLFSSL_MAX_SUITE_SZ);
```
[the remark this answers is not in the document]
Comment (SparkiDev) [reply src/tls13.c:4761]: Put the check into ClientHello parsing.

File: src/tls13.c [hunk 3]
```diff
@@ -4755,16 +4755,34 @@ static void RefineSuites(WOLFSSL* ssl, Suites* peerSuites)
...
 
-    for (i = 0; i < ssl->suites->suiteSz; i += 2) {
+    if (!ssl->options.useClientOrder) {
+        /* Server order refining. */
+        for (i = 0; i < ssl->suites->suiteSz; i += 2) {
+            for (j = 0; j < peerSuites->suiteSz; j += 2) {
+                if (ssl->suites->suites[i+0] == peerSuites->suites[j+0] &&
+                    ssl->suites->suites[i+1] == peerSuites->suites[j+1]) {
+                    suites[suiteSz++] = peerSuites->suites[j+0];
```
[the remark this answers is not in the document]
Comment (SparkiDev) [reply src/tls13.c:4769]: Now culling duplicates.
Safety check put in.

File: src/tls13.c [hunk 4]
```diff
@@ -4755,16 +4755,34 @@ static void RefineSuites(WOLFSSL* ssl, Suites* peerSuites)
...
+        /* Client order refining. */
         for (j = 0; j < peerSuites->suiteSz; j += 2) {
-            if (ssl->suites->suites[i+0] == peerSuites->suites[j+0] &&
-                ssl->suites->suites[i+1] == peerSuites->suites[j+1]) {
-                suites[suiteSz++] = peerSuites->suites[j+0];
-                suites[suiteSz++] = peerSuites->suites[j+1];
+            for (i = 0; i < ssl->suites->suiteSz; i += 2) {
+                if (ssl->suites->suites[i+0] == peerSuites->suites[j+0] &&
+                    ssl->suites->suites[i+1] == peerSuites->suites[j+1]) {
+                    suites[suiteSz++] = peerSuites->suites[j+0];
```
[the remark this answers is not in the document]
Comment (SparkiDev) [reply src/tls13.c:4782]: As above.

**Review by douzzer — APPROVED** [review]
LGTM -- retested with super-quick-check for good measure.

**Review by JacobBarthelmeh — APPROVED** [review]

**Pull request merged**